scripts/build.py

```diff
@@ -631,25 +734,80 @@ def rebuild_new_files(
         recently_changed_filenames: list(str). List of filenames that were
             recently changed.
         file_hashes: dict(str, str). Dictionary of file hashes.
+        build_tasks: deque(thread). A deque that contains all build tasks queued
+            to be processed.
     """
     for file_name in recently_changed_filenames:
-        print 'Minifying %s' % file_name
         source_file_path = os.path.join(source_path, file_name)
         target_file_path = os.path.join(target_path, file_name)
-        ensure_directory_exists(target_file_path)
-        minify_func(source_file_path, target_file_path, file_hashes, file_name)
+        if not does_directory_tree_exist(target_file_path):
+            os.makedirs(os.path.dirname(target_file_path))
+        task = threading.Thread(target=minify_func, args=(
+            source_file_path, target_file_path, file_hashes, file_name))
+        build_tasks.append(task)
 
 
-def remove_deleted_files(dev_directory, staging_directory):
-    """Walk the staging directory and remove files that have since been removed
-    from the DEV directory.
+def build_directory(source_dir, file_hashes, build_tasks):
+    """Build all files in provided DEV directory if there is no existing staging
+    dir. Otherwise, selectively build recently changed files in
+    said DEV directory and copy to staging dir.
 
     Args:
-        dev_directory: str. Path relative to /oppia where DEV files are located.
+        source_dir: dict(str, str). Directory dict containing directory paths
+        to /DEV dir, staging dir and /build dir .
+        file_hashes: dict(str, str). Dictionary of file hashes.
+        build_tasks: deque(thread). A deque that contains all build tasks queued
+            to be processed.
+    """
+    dev_dir = source_dir.get('dev_dir')
```

Comment:
Is source_dir actually a directory? If not it should be renamed correctly.

Here and elsewhere, why use .get() instead of source_dir['dev_dir']? Latter is better since it throws KeyError if the expected key-value pair doesn't exist.

Reply:
Renamed dict object to `dirname_dict` and used array notation instead of .get() method.
Thanks!